Consolidate ert & everest interactions with plugins

yngve-sk opened the pull request:
**Issue**
Resolves https://github.com/equinor/ert/issues/11673


**Approach**
* Resolve circular dependency by moving workflows to ert.config instead of ert.plugins, as ert.config can no longer import ert.plugins, but ert.plugins needs to import config to parse/read the site config.
* Pass in a typed object with plugin context to pydantic validators
* ERT config.with_plugins now takes in the context from the ErtPluginContext(), removing any direct interactions with the ErtPluginManager. The custom error handling, initialization and heavy use of staticmethods in ErtConfig makes it unsuitable for (without a big refactor) taking in the context via pydantic in sub-validators.
* RunModels receive the necessary context in through pydantic validation context

(Screenshot of new behavior in GUI if applicable)


- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Added appropriate release note label
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Make sure unit tests pass locally after every commit (`git rebase -i main
      --exec 'just rapid-tests'`)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Add backport label to latest release (format: 'backport release-branch-name')

<!--
Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).
-->

yngve-sk commented on tests/ert/unit_tests/plugins/dummy_plugins.py:
```diff
@@ -18,27 +18,27 @@ def forward_model_configuration():
 @plugin(name="dummy")
 def ecl100_config_path():
-    return "/dummy/path/ecl100_config.yml"
+    return "dummy/path/ecl100_config.yml"
```
The tests will now actually create an ertconfig from the site config lines, so they require files to exist. Leading slash would point to root dir on the computer running the tests

yngve-sk commented on src/everest/config/simulator_config.py:
```diff
@@ -113,7 +113,7 @@ def default_local_queue(cls, v: Any, info: ValidationInfo) -> Any:
             if info.context:
-                options = info.context.get("queue_system")
+                options = info.context.queue_options
             return options or LocalQueueOptions(max_running=8)
```
Hm, I think it can still be None? queue options are set here:
```
queue_options=site_config.queue_config.queue_options
            if has_site_config
            else None,
```
and typing there does include optional

yngve-sk commented on tests/ert/ui_tests/cli/run_cli.py:
```diff
@@ -3,9 +3,9 @@
+from ert.plugins.plugin_manager import ErtPluginContext, ErtRuntimePlugins
 
-PM = ErtPluginManager()
+RUNTIME_PLUGINS = ErtPluginContext().__enter__()  # noqa: PLC2801
```
Made it slightly better, will see very soon if consolidating the `init_context` / `with ErtPluginContext(...)` / `ErtRuntimePlugins` concepts will open up for doing this in a cleaner way

oyvindeide commented on src/ert/config/ert_config.py:
```diff
@@ -1054,6 +1014,10 @@ def from_dict(cls, config_dict: ConfigDict) -> Self:
+
+        # Insert env vars from plugins/site config
+        for key, val in cls.ENV_VARS.items():
+            env_vars[key] = val  # Q: Substitute these? From site
```
No, I think warn if duplicate, then take user value.

oyvindeide commented on src/ert/config/queue_config.py:
```diff
@@ -314,19 +343,52 @@ class QueueConfig(BaseModel):
+    def from_dict(
+        cls,
+        config_dict: ConfigDict,
+        site_queue_options: QueueOptions | None = None,
```
This is easier than changing to: `config_content_to_dict`?

yngve-sk replied:
I'm assuming you mean `ErtConfig.from_dict` where this is being called? I think it is easier to do it here, there are a lot of edge cases packed in here and the logic currently (still in this PR) reads complex. So I think we do need one place where we have the
1. list of lines in ertconfig that sets a queue option (some prefixed by `QUEUE_OPTION <SYSTEM>`, some are just not prefixed)
2. the options set in the site config

the order in which we want things is something like user setting > site config setting > default setting, and `QUEUE_OPTION <O>` settings should override global settings `<O>`, and there are some loggings given when we override things, I also added a logging for when a user config overrides site config as seen in the revised `test_that_overwriting_QUEUE_OPTIONS_warns`.

oyvindeide replied:
I was thinking that this function would be:
```python
def config_content_to_dict(config: ConfigDict) -> unpack[QueueConfig]:
    return extract_queue_options_from_config_dict
```

oyvindeide commented on src/ert/plugins/plugin_manager.py:
```diff
@@ -182,6 +186,24 @@ def get_forward_model_configuration(self) -> dict[str, dict[str, Any]]:
+                    f"Only one site configuration is allowed, got {plugin_names}"
+                )
+            return ErtRuntimePlugins.model_validate(plugin_response[0].data)
+        except AttributeError:
```
What is this from?

oyvindeide commented on tests/everest/conftest.py:
```diff
@@ -287,27 +285,12 @@ def no_plugins():
-
-
-@pytest.fixture(autouse=True)
-def no_site_config():
```
How will testing in an environment with a site config work now?

yngve-sk replied:
Can change the fixture a bit, but basically we just need to mock-override the `RuntimePlugins` class to return `None` of a default empty set of plugins, for the actual site config environment to never reach the test. Did this in some Everest tests already, ref 
```
with mock.patch(
        "ert.plugins.plugin_manager.ErtRuntimePlugins",
        return_value=ErtRuntimePlugins(
            environment_variables={"HOW_MANY_CPU": "<NUM_CPU>"}
        ),
    ):
```
could just return an empty instead

oyvindeide commented on src/ert/config/ert_config.py:
```diff
@@ -668,10 +663,13 @@ class ErtConfig(BaseModel):
-    PREINSTALLED_WORKFLOWS: ClassVar[dict[str, ErtScriptWorkflow]] = {}
+    PREINSTALLED_WORKFLOWS: ClassVar[dict[str, WorkflowJob]] = {}
     ENV_PR_FM_STEP: ClassVar[dict[str, dict[str, Any]]] = {}
     ACTIVATE_SCRIPT: ClassVar[str | None] = None
```
Will this be configured twice?

eivindjahren commented on src/ert/config/ert_config.py:
```diff
@@ -780,40 +776,19 @@ def __eq__(self, other: object) -> bool:
-                pm.get_ertscript_workflows().get_workflows()
-                | pm.get_legacy_ertscript_workflows().get_workflows()
+            ] = runtime_plugins.installed_forward_model_steps
+            PREINSTALLED_WORKFLOWS = dict(runtime_plugins.installed_workflow_jobs)
```
There seems to be some changes in this commit that doesn't belong to it. specifically `with_plugins` has a lot of changes not related to resolving circular dependencies.

eivindjahren commented on docs/ert/getting_started/howto/plugin_system.rst:
```diff
@@ -233,7 +233,7 @@ Full example:
 
    @ert.plugin(name="my_plugin")
-    def ertscript_workflow(config: ert.WorkflowConfigs):
+    def ertscript_workflow(config: ert.config.WorkflowConfigs):
```
This we really do not want. A plugin should never have to import anything from `ert.config`.

yngve-sk replied:
fixed

eivindjahren reviewed (changes requested):
There is a change to the plugin system docs that we really need to avoid. Other than that, more granular commits would make this a lot easier to review. Mostly the first commit seems to contain changes that are at odds with the commit message title and body.

eivindjahren commented on src/ert/config/ert_config.py:
```diff
@@ -808,10 +810,9 @@ class ErtConfigWithPlugins(ErtConfig):
                 Mapping[str, ForwardModelStep]
             ] = preinstalled_fm_steps
-            PREINSTALLED_WORKFLOWS = (
+            PREINSTALLED_WORKFLOWS = dict(
```
There are still some changes here that are unrelated to moving workflows into the config module

eivindjahren reviewed (dismissed):
Found some small issues that should be resolved. I will look closer at the actual changes to see if there are bigger suggestions.

eivindjahren commented on src/ert/config/ert_config.py:
```diff
@@ -780,41 +776,19 @@ def __eq__(self, other: object) -> bool:
-                pm.get_ertscript_workflows().get_workflows()
-            ) | dict(pm.get_legacy_ertscript_workflows().get_workflows())
-            ENV_PR_FM_STEP: ClassVar[dict[str, dict[str, Any]]] = env_pr_fm_step_
-            ACTIVATE_SCRIPT = pm.activate_script()
```
The removal of the `activate_script` hook seems a rather independent part of what is being done here. It would simplify understanding what has changed by having a separate commit for that change alone.

eivindjahren commented on src/ert/config/workflow_job.py:
```diff
@@ -143,11 +150,36 @@ class ErtScriptWorkflow(_WorkflowJob):
     """
 
-    ert_script: type[ErtScript] = None  # type: ignore
+    ert_script: builtins.type[ErtScript] = None  # type: ignore
```
Making `ert_script` serializable seems to not be related to the change here, and could probably easily be in its own commit.

eivindjahren commented on tests/ert/ui_tests/cli/test_cli.py:
```diff
@@ -684,6 +684,7 @@ def test_that_prior_is_not_overwritten_in_ensemble_experiment(
             "COEFFS", np.arange(num_realizations)
         )
         np.testing.assert_array_equal(parameter_values, prior_values)
+
```
This extra line should probably not be here.

eivindjahren commented on tests/ert/unit_tests/config/test_ert_config.py:
```diff
@@ -324,7 +315,7 @@ def run(self, *args):
         dedent(
             """
-            from ert.plugins import ErtPlugin
+            from ert.config import ErtPlugin
```
This change should be part of the previous commit. Please make sure the tests pass on all commits.

eivindjahren commented on src/ert/config/ert_config.py:
```diff
@@ -1396,7 +1307,7 @@ def uppercase_subkeys_and_stringify_subvalues(
 
 
-def _forward_model_step_from_config_contents(
+def forward_model_step_from_config_contents(
```
Please put the rename of this function into its own commit.

eivindjahren commented:
A couple of extra things we should probably test for:
* What if multiple plugins provide the `site_configurations` hook.
* There is a test for warning for overriding queue options, but we should have the same for environment variables, workflow jobs and forward model steps. It is also important that we assert which takes precedence.

yngve-sk commented:
> A couple of extra things we should probably test for:
> 
> * What if multiple plugins provide the `site_configurations` hook.

Added test for this

> * There is a test for warning for overriding queue options, but we should have the same for environment variables, workflow jobs and forward model steps. It is also important that we assert which takes precedence.

Added for overrides of envvars. 

For workflows and forward model steps: They overriding is handled to some extent within the plugin manager, so all we should (imo) explicitly test is that user settings override "installed" (from plugins, not only limited to site config) ones.

**Workflows:** ref `test_that_adding_a_workflow_twice_warns`,  `test_that_ert_warns_on_duplicate_workflow_jobs`. Legacy workflows are not overriding, there new additions are skipped, this is asserted in `test_workflow_config_duplicate_log_message`. Suggest to not touch/revise this in this PR.

**Forward model steps:** Already tested to some extent, `test_that_installing_two_forward_model_steps_with_the_same_name_warn`, `test_that_installing_two_forward_model_steps_with_the_same_name_warn_with_dir`. Wrote in a clause to give a warning for overrides of site config, as well as a test for precedence. Should be rewritten more but this should be deferred to a future PR.

yngve-sk commented on tests/ert/unit_tests/config/test_queue_config.py:
```diff
@@ -176,36 +176,108 @@ def test_invalid_realization_memory(invalid_memory_spec: str, error_message: str
+    )
+
+
+def test_that_user_given_queue_settings_overwrites_site_config_(
```
fixed

berland commented:
> > No commit message mention anything about what is the breaking change. Can that be made explicit?
> 
> fixed, does this look OK?
> 
> ```
> Rewrite site config to not go through ErtConfig
> 
> Take in site-wide configurations of ERT through
> typed python object instead of ERT config lines.
> 
> This is a non major version bump,
> backward-compatible breaking change, and will
> not work with older specifications of
> site-configurations that are intended to append
> lines to the ERT config. Newer configs must
> instead provide their plugins through the
> site_configurations hook.
> ```
The text says "non major", but you maybe mean "major"? But how we version Ert does not need to go into the commit message, it is more like the versioning is the result of the commits. So the important bit is to say that this commit is breaking, and what it is breaking.

yngve-sk commented:
> > > No commit message mention anything about what is the breaking change. Can that be made explicit?
> > 
> > 
> > fixed, does this look OK?
> > ```
> > Rewrite site config to not go through ErtConfig
> > 
> > Take in site-wide configurations of ERT through
> > typed python object instead of ERT config lines.
> > 
> > This is a non major version bump,
> > backward-compatible breaking change, and will
> > not work with older specifications of
> > site-configurations that are intended to append
> > lines to the ERT config. Newer configs must
> > instead provide their plugins through the
> > site_configurations hook.
> > ```
> 
> The text says "non major", but you maybe mean "major"? But how we version Ert does not need to go into the commit message, it is more like the versioning is the result of the commits. So the important bit is to say that this commit is breaking, and what it is breaking.

oops, the "non" was accidental indeed. Reworded to be along the lines of  "this is a breaking change":

```
This is a breaking change that requires ERT
plugins to be provided through the
site_configurations hook.
```

oyvindeide reviewed (approved):
LGTM!
